File: apps/web/src/types/problem.ts
import { z } from 'zod';

// ============================================================================
// Base Types
// ============================================================================

export const ExampleSchema = z.object({
    input: z.array(z.unknown()),
    output: z.unknown(),
    explanation: z.string().optional(),
});

export const ComplexitySchema = z.object({
    time: z.string(),
    space: z.string(),
});

export const APIParamSchema = z.object({
    name: z.string(),
    type: z.string(),
    desc: z.string(),
});

export const APIParamReturnSchema = z.object({
    type: z.string(),
    desc: z.string(),
});

export const APISignatureSchema = z.object({
    function_name: z.string(),
    signature: z.string(),
    params: z.array(APIParamSchema),
    returns: APIParamReturnSchema,
});

export const ProblemMetadataSchema = z.object({
    title: z.string(),
    statement: z.string(),
    constraints: z.array(z.string()),
    examples: z.array(ExampleSchema),
    edge_cases: z.array(z.string()),
});

export const SolutionOutlineSchema = z.object({
    approach: z.string(),
    complexity: ComplexitySchema,
    code: z.string(),
});

export const TestSuiteSchema = z.object({
    public: z.array(ExampleSchema),
    hidden: z.array(ExampleSchema),
});

export const WorkspaceFileSchema = z.object({
    code: z.string(),
    hidden: z.boolean().optional(),
});

export const WorkspaceTemplateSchema = z.object({
    entry: z.string(),
    files: z.record(WorkspaceFileSchema),
    dependencies: z.record(z.string()).optional(),
    dev_dependencies: z.record(z.string()).optional(),
    template: z.string().optional(),
    environment: z.string().optional(),
});

export const ProblemPackSchema = z.object({
    problem: ProblemMetadataSchema,
    api: APISignatureSchema,
    time_estimate_minutes: z.number(),
    hint: z.string(),
    solutions: z.array(SolutionOutlineSchema),
    tests: TestSuiteSchema,
    macro_category: z.enum(['dsa', 'frontend', 'system-design']),
    workspace_template: WorkspaceTemplateSchema.optional(),
});

export const RunResultSchema = z.object({
    test_id: z.string(),
    status: z.string(),
    time_ms: z.number(),
    stdout: z.string(),
    stderr: z.string(),
});

export const AttemptSchema = z.object({
    id: z.string(),
    problem_id: z.string(),
    user_id: z.string(),
    lang: z.string(),
    started_at: z.number(),
    ended_at: z.number(),
    hint_used: z.boolean(),
    pass_count: z.number(),
    fail_count: z.number(),
    duration_ms: z.number(),
});

// ============================================================================
// Type Exports
// ============================================================================

export type Example = z.infer<typeof ExampleSchema>;
export type Complexity = z.infer<typeof ComplexitySchema>;
export type APIParam = z.infer<typeof APIParamSchema>;
export type APIParamReturn = z.infer<typeof APIParamReturnSchema>;
export type APISignature = z.infer<typeof APISignatureSchema>;
export type ProblemMetadata = z.infer<typeof ProblemMetadataSchema>;
export type SolutionOutline = z.infer<typeof SolutionOutlineSchema>;
export type TestSuite = z.infer<typeof TestSuiteSchema>;
export type ProblemPack = z.infer<typeof ProblemPackSchema>;
export type WorkspaceTemplate = z.infer<typeof WorkspaceTemplateSchema>;
export type WorkspaceFile = z.infer<typeof WorkspaceFileSchema>;
export type RunResult = z.infer<typeof RunResultSchema>;
export type Attempt = z.infer<typeof AttemptSchema>;

// ============================================================================
// UI-specific Types
// ============================================================================

export type MacroCategory = 'dsa' | 'frontend' | 'system-design';

export type DsaCategory =
    | 'arrays'
    | 'bfs-dfs'
    | 'maps-sets'
    | 'dp'
    | 'graphs'
    | 'strings'
    | 'math'
    | 'heaps'
    | 'two-pointers';

export type FrontendCategory =
    | 'react-components'
    | 'css-layouts'
    | 'accessibility'
    | 'state-management'
    | 'performance'
    | 'forms-validation';

export type SystemDesignCategory =
    | 'scalability'
    | 'databases'
    | 'caching'
    | 'load-balancing'
    | 'microservices'
    | 'api-design';

export type Category = DsaCategory | FrontendCategory | SystemDesignCategory;
export type Difficulty = 'easy' | 'medium' | 'hard';
export type Provider = 'openai' | 'grok';
export type TestVisibility = 'public' | 'hidden';
export type TestStatus = 'pass' | 'fail' | 'timeout' | 'error';

// ============================================================================
// Test Result UI Type
// ============================================================================

export interface TestResult {
    test_id: string;
    status: TestStatus;
    time_ms: number;
    stdout?: string;
    stderr?: string;
    expected?: unknown;
    actual?: unknown;
}

File: backend/internal/domain/problem.go
package domain

// Example represents one illustrative or test case for a generated problem.
type Example struct {
	Input       []any  `json:"input"`
	Output      any    `json:"output"`
	Explanation string `json:"explanation,omitempty"`
}

// ProblemMetadata contains the human-readable statement and constraints.
type ProblemMetadata struct {
	Title       string    `json:"title"`
	Statement   string    `json:"statement"`
	Constraints []string  `json:"constraints"`
	Examples    []Example `json:"examples"`
	EdgeCases   []string  `json:"edge_cases"`
}

// APISignature describes the expected function signature for the solution code.
type APISignature struct {
	FunctionName string         `json:"function_name"`
	Signature    string         `json:"signature"`
	Params       []APIParam     `json:"params"`
	Returns      APIParamReturn `json:"returns"`
}

// APIParam captures an argument description.
type APIParam struct {
	Name string `json:"name"`
	Type string `json:"type"`
	Desc string `json:"desc"`
}

// APIParamReturn captures the return description.
type APIParamReturn struct {
	Type string `json:"type"`
	Desc string `json:"desc"`
}

// SolutionOutline represents an example solution with complexity metadata.
type SolutionOutline struct {
	Approach   string     `json:"approach"`
	Complexity Complexity `json:"complexity"`
	Code       string     `json:"code"`
}

// Complexity describes Big-O characteristics of a solution.
type Complexity struct {
	Time  string `json:"time"`
	Space string `json:"space"`
}

// TestSuite groups public and hidden tests for a generated problem.
type TestSuite struct {
	Public []Example `json:"public"`
	Hidden []Example `json:"hidden"`
}

// WorkspaceFile represents a starter file within the interactive sandbox.
type WorkspaceFile struct {
	Code   string `json:"code"`
	Hidden bool   `json:"hidden,omitempty"`
}

// WorkspaceTemplate defines the initial multi-file sandbox configuration.
type WorkspaceTemplate struct {
	Entry           string                   `json:"entry"`
	Files           map[string]WorkspaceFile `json:"files"`
	Dependencies    map[string]string        `json:"dependencies,omitempty"`
	DevDependencies map[string]string        `json:"dev_dependencies,omitempty"`
	Template        string                   `json:"template,omitempty"`
	Environment     string                   `json:"environment,omitempty"`
}

// ProblemPack is the full payload returned by the LLM broker.
type ProblemPack struct {
	Problem           ProblemMetadata    `json:"problem"`
	API               APISignature       `json:"api"`
	TimeEstimateMins  int                `json:"time_estimate_minutes"`
	Hint              string             `json:"hint"`
	Solutions         []SolutionOutline  `json:"solutions"`
	Tests             TestSuite          `json:"tests"`
	MacroCategory     string             `json:"macro_category"`
	WorkspaceTemplate *WorkspaceTemplate `json:"workspace_template,omitempty"`
}

// Attempt captures stored attempt metadata.
type Attempt struct {
	ID         string `json:"id"`
	ProblemID  string `json:"problem_id"`
	UserID     string `json:"user_id"`
	Language   string `json:"lang"`
	StartedAt  int64  `json:"started_at"`
	EndedAt    int64  `json:"ended_at"`
	HintUsed   bool   `json:"hint_used"`
	PassCount  int    `json:"pass_count"`
	FailCount  int    `json:"fail_count"`
	DurationMS int64  `json:"duration_ms"`
}

// RunResult captures output from executing a single test case.
type RunResult struct {
	TestID string `json:"test_id"`
	Status string `json:"status"`
	TimeMS int64  `json:"time_ms"`
	Stdout string `json:"stdout"`
	Stderr string `json:"stderr"`
}

// RunSummary groups multiple run results.
type RunSummary struct {
	AttemptID string      `json:"attempt_id"`
	Results   []RunResult `json:"results"`
}

// SubmissionSummary contains outcome metrics after running hidden tests.
type SubmissionSummary struct {
	AttemptID     string      `json:"attempt_id"`
	Passed        bool        `json:"passed"`
	RuntimeMS     int64       `json:"runtime_ms"`
	Operations    int64       `json:"operations"`
	HiddenResults []RunResult `json:"hidden_results"`
}
